cocos/ui/UIWebViewImpl-android.cpp

```diff
@@ -401,45 +404,73 @@ namespace cocos2d {
                 setScalesPageToFitJNI(_viewTag, scalesPageToFit);
             }
 
+            // executed from Java's thread (we should invoke callback on cocos thread)
             bool WebViewImpl::shouldStartLoading(const int viewTag, const std::string &url) {
-                auto it = s_WebViewImpls.find(viewTag);
-                if (it != s_WebViewImpls.end()) {
-                    auto webView = s_WebViewImpls[viewTag]->_webView;
-                    if (webView->_onShouldStartLoading) {
-                        return webView->_onShouldStartLoading(webView, url);
+
+                std::mutex m;
+                std::condition_variable cv;
+                bool finished = false;
+                bool allowLoad = true;
+
+                Director::getInstance()->getScheduler()->performFunctionInCocosThread([&]{
+                    auto it = s_WebViewImpls.find(viewTag);
+                    if (it != s_WebViewImpls.end()) {
+                        auto webView = it->second->_webView;
+                        if (webView->_onShouldStartLoading) {
+                            allowLoad = webView->_onShouldStartLoading(webView, url);
+                        }
                     }
-                }
-                return true;
+
+                    m.lock();
+                    finished = true;
+                    m.unlock();
+                    cv.notify_one();
+                });
+
+                // wait for result from cocos thread
+                std::unique_lock<std::mutex> lk(m);
```

Comment:
This may block Android UI thread if the developers do some heavy things in Cocos thread.
And I have added some comments to the `shouldStartLoading` declaration. Don't do any OpenGL related operations here. So if you improve this method, I think it's better to also improve the comments. Thanks.